smf: improve sibling transitions speed

vlad-kulikov opened the pull request:
This PR introduces `CONFIG_SMF_FAR_TRANSITION` to optimize transitions
between distant hierarchical states (classified as "far" when the depth
difference between source and destination is >= 3). When enabled, SMF
can use a hybrid strategy: near transitions use the classic LCA, and far
transitions use a dedicated far-LCA routine tuned for deep hierarchies.
Public APIs are unchanged.

* New variable ,`int32_t transition`, is added to `struct smf_ctx` (built only when 
`CONFIG_SMF_FAR_TRANSITION` is enabled).
A state can request the far-LCA algorithm for the next transition by setting
`ctx->transition = SMF_TRANSITION_FAR`.
During `smf_set_state()`, the transition type is read. If it is FAR, `get_lca_far()`
is used. The transition type is then reset to NEAR so subsequent transitions
use the classic LCA unless explicitly requested again.
 
* New `get_depth()` helper was added to calculate the depth of a state from root
(built only when `CONFIG_SMF_FAR_TRANSITION` is enabled):
`static inline uint8_t get_depth(const struct smf_state *state)`
 
* New enumeration is introduced in order to describe the type of transition:
```
  /**
 * @brief enum for choosing which transition function suits best with smf_set_state
 */
enum smf_transition_distance {
	SMF_TRANSITION_NEAR,
	SMF_TRANSITION_FAR
};
```
 
* Added a new ztest for testing benchmark on different boards. 
 
Micro-benchmark setup
- Board: NUCLEO-F746ZG (STM32F746, ~192 MHz)
- 200000 iterations per pair
- HSM topology (R is root):
      R
    /   \
   A     B
  / \   / \
 C  D  E  F
/ \        \
G  H        J
- Entry/exit are NULL to measure raw transition overhead.
- Results are cycles/transition; % is relative to baseline.

Baseline (CONFIG_SMF_FAR_TRANSITION=n)
  G<->G: 229 cycles
  C->G : 308 cycles
  A->H : 423 cycles
  G->H : 458 cycles
  G->D : 606 cycles
  D->E : 700 cycles
  J->D : 856 cycles
  J->G : 1025 cycles

Enabled but not used (CONFIG_SMF_FAR_TRANSITION=y; far-LCA never taken)
  G<->G: 262 vs 229 cycles  = +33 (+14.40%)
  C->G : 347 vs 308 cycles = +39 (+12.66%)
  A->H : 455 vs 423 cycles = +32 (+7.57%)
  G->H : 494 vs 458 cycles = +36 (+7.86%)
  G->D : 614 vs 606 cycles =  +8 (+1.32%)
  D->E : 703 vs 700 cycles =  +3 (+0.43%)
  J->D : 819 vs 856 cycles =  -37 (-4.32%)  improvement
  J->G : 957 vs 1025 cycles =  -68 (-6.63%)  improvement

Hybrid (only far pairs use far-LCA)
  G<->G: 233 vs 229 cycles =  +4 (+1.75%)
  C->G : 352 vs 308 cycles = +44 (+14.29%)
  A->H : 489 vs 423 cycles = +66 (+15.60%)
  G->H : 552 vs 458 cycles = +94 (+20.52%)
  G->D : 646 vs 606 cycles = +40 (+6.60%)
  D->E : 739 vs 700 cycles = +39 (+5.57%)
  J->D : 743 vs 856 cycles = -113 (-13.20%) improvement
  J->G : 817 vs 1025 cycles = -208 (-20.29%) improvement

All-far (force far-LCA for all transitions) -- not proposed
  G<->G: 378 vs 229 cycles = +149 (+65.07%)
  C->G : 430 vs 308 cycles = +122 (+39.61%)
  A->H : 612 vs 423 cycles = +189 (+44.68%)
  G->H : 483 vs 458 cycles =  +25 (+5.46%)
  G->D : 594 vs 606 cycles =  -12 (-1.98%)  improvement
  D->E : 624 vs 700 cycles =  -76 (-10.86%) improvement
  J->D : 729 vs 856 cycles = -127 (-14.84%) improvement
  J->G : 802 vs 1025 cycles = -223 (-21.76%) improvement

Summary
- Hybrid mode improves far transitions by ~13–20% on this setup while
  keeping default behavior unchanged when the option is off.
- Enabling the option introduces small overheads on near/mid paths if
  far-LCA is not taken; therefore default remains 'n'.
- Greater depth differences generally yield larger speedups because
   the far-LCA routine runs in O(h) time, while the classic ancestor-scan
   approach is O(h^2), in worst case. Here h is the number of steps from
   the deeper nodes to the LCA. As h grows, the far-LCA avoids repeated
   re-traversals and wins proportionally more.
 - Calculated use of far-LCA routine, can improve the speed of transitions
    in the HSM drastically.

vlad-kulikov commented:
> What would be nice to know is how often these types of transitions occur. The speedup for far transitions is nice, but most transitions in state machines I use are within the same nested sub-machione, with a single transition to the next sub-machine when the work performed by that sub-machine is complete.
> 
> What are the use cases from jumping from sub-machine to sub-machine frequently enough that the tradeoff in near jumps is worth it?

Hi @glenn-andrews you are absolutely right. Most apps do mostly near hops. That’s why this is fully optional (default off). No cost if unused, it’s there for time-constrained cases.

vlad-kulikov commented:
I see your point @glenn-andrews @keith-zephyr . 
Instead of pushing a far-transition speedup, I added an optional sibling
fast-path: CONFIG_SMF_LCA_SIBLING_SHORTCUT. 
It detects the common case where `src->parent == dst->parent` and 
skips the general LCA walk, keeping sibling hops effectively constant-time regardless of nesting depth.

``` 
         R
       /   \
      A     B
     / \   / \
    C  D  E  F
   / \        \
  G  H        J

Micro-bench (cycles/transition, ns):
(A→B) 475 → 320 (~33% faster) — 2473 ns → 1666 ns
(C→D) 507 → 320 (~37% faster) — 2640 ns → 1666 ns
(G→H) 545 → 320 (~41% faster) — 2838 ns → 1666 ns

```
The baseline get_lca_of() cost grows with deeper nesting even for sibling→sibling, whereas the shortcut stays flat.

vlad-kulikov commented:
> The current PR has only the sibling transition. I removed the far transition.

Please review this PR, @keith-zephyr  @glenn-andrews .

glenn-andrews commented on include/zephyr/smf.h:
```diff
@@ -65,6 +65,14 @@ enum smf_state_result {
+/**
+ * @brief enum for enabling/disabling get_lca_sibling_of function
+ */
+enum smf_lca_sibling_status {
```
Can this be put in `smf.c`? I don't see it needing global scope.

glenn-andrews reviewed (changes requested):
I think you stripped out too much code. Can you double check?

vlad-kulikov committed:
smf: improve sibling transitions speed

Sibling transitions are now detected in smf_set_state() when source and
destination share the same parent, removing the need for states to flag
the sibling transition explicitly.

Restore include/zephyr/smf.h and lib/smf/Kconfig to match main so this
change only carries the functional update in lib/smf/smf.c, and add a
micro-benchmark test. For accurate statistics, run the test on a board.

Micro-benchmark setup
- Board: NUCLEO-F746ZG (STM32F746, ~192 MHz)
- 200000 iterations per state transition
- Simple HSM (letters are nodes; R is root):

=== SMF transition micro-benchmark ===
        R
      /   \
     A     B
    / \   / \
   C   D E   F
  / \         \
 G   H         J

Current smf_set_state:
======================
Sibling Transitions
(A->B) : 459 cycles/transition (2390 ns)
(C->D) : 482 cycles/transition (2510 ns)
(G->H) : 522 cycles/transition (2718 ns)

Other Transitions
(G<->G) : 237 cycles/transition (1234 ns)
(C->G) : 343 cycles/transition (1786 ns)
(A->H) : 452 cycles/transition (2354 ns)
(G->D) : 651 cycles/transition (3390 ns)
(D->E) : 752 cycles/transition (3916 ns)
(J->D) : 893 cycles/transition (4651 ns)
(J->G) : 1077 cycles/transition (5609 ns)

New smf_set_state:
==================
Sibling Transitions
(A->B) : 356 cycles/transition (1854 ns)(22% faster)
(C->D) : 356 cycles/transition (1854 ns)(26% faster)
(G->H) : 356 cycles/transition (1854 ns)(32% faster)

Other Transitions
(G<->G) : 258 cycles/transition (1343 ns)(9% slower)
(C->G) : 356 cycles/transition (1854 ns)(4% slower)
(A->H) : 464 cycles/transition (2416 ns)(3% slower)
(G->D) : 707 cycles/transition (3682 ns)(9% slower)
(D->E) : 797 cycles/transition (4151 ns)(6% slower)
(J->D) : 970 cycles/transition (5052 ns)(9% slower)
(J->G) : 1157 cycles/transition (6026 ns)(8% slower)

This change makes sibling transitions
deterministic and cheaper (356 cycles on nsim),
at the cost of a small increase for deeper/LCA
transitions (~7–9%). This is a net win for state machines
that mostly hop between siblings.

Signed-off-by: Vladislav Kulikov <vlad_kulikov_c@pm.me>

vlad-kulikov commented on lib/smf/smf.c:
```diff
@@ -302,6 +302,16 @@ void smf_set_state(struct smf_ctx *const ctx, const struct smf_state *new_state)
 	const struct smf_state *topmost;
 
+#ifdef CONFIG_SMF_LCA_SIBLING_SHORTCUT
+	if (ctx->sibling == SMF_LCA_SIBLING_ENABLE) {
```
This code is fine. Before `calling smf_set_state()`, a state would enable the sibling transition with `SMF_LCA_SIBLING_ENABLE`. Inside `smf_set_state(),` after the sibling LCA was handled,
`ctx->sibling` was reset to `SMF_LCA_SIBLING_DISABLE`. That left the choice of using the faster sibling transition to the state, which meant the user was responsible for setting a correct sibling transition and SMF needed less logic. As a result, transitions were faster than in the new version I added, which is expected.

glenn-andrews reviewed (changes requested):
Fix the tree diagram.

glenn-andrews reviewed (dismissed):
Thank you.

I think the next thing might be pull the flat transition and sibling transition code into a single function and then it should go even faster since you don't need to iterate through `smf_execute_all_exit_actions()` and `smf_execute_all_entry_actions()`

vlad-kulikov commented:
Please review this PR, @keith-zephyr .

keith-zephyr commented on lib/smf/smf.c:
```diff
@@ -300,6 +300,13 @@ void smf_set_state(struct smf_ctx *const ctx, const struct smf_state *new_state)
+
+		goto lca_sibling;
+	}
+
```
You don' t need the goto.


```
	if (ctx->executing != new_state && ctx->executing->parent == new_state->parent) {
		/* Optimize sibling transitions (different states under same parent) */
		topmost = ctx->executing->parent;
	} else if (is_descendant_of(ctx->executing, new_state)) {
...
```

vlad-kulikov replied:
You are right, this logic is connected to my previous changes. I will change it, thank you.
